### PR TITLE
fix "Can't use synchronousEvent inside a listener"-Issue

### PR DESCRIPTION
**Pull request by tehsphinx**
New pull request to get rid of fork commit:

Events are really synchronous now, waiting for the all listeners for finish before moving on.

Just to underline the importance of fixing this issue here a simple example that does not work with the current code:
- Create a application that minimizes to tray on startup
- Add a menu item to the tray that shows the main window

```
[]*astilectron.MenuItemOptions{
		{
			Label:   astilectron.PtrStr("Show"),
			OnClick: func(e astilectron.Event) (deleteListener bool) {
				// this call will make the dispatcher hang and block it from executing any more events
				window.Show()
				return
			},
		},
}
```

### CODE REVIEW

File: dispatcher.go [hunk 1]
```diff
@@ -68,19 +73,28 @@ func (d *Dispatcher) delListener(targetID, eventName string, id int) {
 
 // Dispatch dispatches an event
 func (d *Dispatcher) Dispatch(e Event) {
-	d.c <- e
+	done := make(chan bool)
+	d.c <- dispatchCmd{
+		event: e,
+		done:  done,
+	}
+	<-done
```
Comment (tehsphinx) [inline dispatcher.go]: `Dispatch` will wait for the `start` routine to say it is done

File: dispatcher.go [hunk 2]
```diff
@@ -68,19 +73,28 @@ func (d *Dispatcher) delListener(targetID, eventName string, id int) {
...
+		case cmd := <-d.c:
+			// needed so dispatches of events triggered in the listeners can be received without blocking
+			go func() {
+				for id, l := range d.listeners(cmd.event.TargetID, cmd.event.Name) {
+					if l(cmd.event) {
+						d.delListener(cmd.event.TargetID, cmd.event.Name, id)
+					}
 				}
-			}
+				cmd.done <- true
```
Comment (tehsphinx) [inline dispatcher.go]: With the done channel the `start` routine signals it is finished executing all listeners.

**Comment by tehsphinx** [comment]
https://github.com/asticode/go-astilectron/issues/49

**Comment by asticode** [comment]
@tehsphinx I finally had time to go through this PR, sorry for the delay.

The more I look at the problem you're trying to solve and the more I tend to agree with you regarding executing listeners in a go routine.

I'm even at a point where I find it useless to block in the `Dispatch` function (which was maybe your first thought. if that's the case sorry for that :D).

As far as I'm concern, **events should be dispatched asynchronously**

If we need to make sure a function blocks until it has received a specific event, we use the `synchronousEvent` function like it's used all over this library.

Therefore this PR would "only" add the execution of listeners in the dispatcher's `start` function in a GO routine.

Is there something I'm missing out or do you agree with my train of thoughts?

Cheers

**Comment by tehsphinx** [comment]
Sorry for the wait. I had been preoccupied with other things at work.

> Therefore this PR would "only" add the execution of listeners in the dispatcher's start function in a GO routine.

Basically yes. It looks a bit more complex because I tried to keep it synchronous as much as possible. The routine calling an event will still wait for the event to finish -- but not for events called by events. Also other events can fire asynchronously from other go routines:

Sample:
- Goroutine 1 fires Event 1
- Dispatcher starts execution of Event 1 in goroutine executing all listeners SYNCHRONOUSLY
- Goroutine 2 fires Event 2
- Dispatcher starts execution of Event 2 in goroutine executing all listeners SYNCHRONOUSLY
- Event 1 is done, Goroutine 1 continues
- Event 2 is done, Goroutine 2 continues

There is another key change: At the same time it also starts events coming back in from electron in a go routine. If we don't do that the reader is blocked and cannot receive more events from electron. Here I see no problem at all, since the reader would not do anything after executing the event except read the next event.

**Comment by asticode** [comment]
First off I've just unexported the dispatcher because it didn't make sense to export the entity responsible to dispatch **internal** events. Therefore the `dispatch` function is only called internally by the `reader` and `*Astilectron` (to dispatch "app has crashed", "app has stopped" events). This created conflicts with your PR and I apologize for that. Can you rebase your PR on master?

What that also means is that I don't think you need to keep the `dispatch` synchronous anymore. No need to execute the reader's dispatch in a go routine either. That way all events are dispatched **asynchronously** which I think I should have done from the get go. Therefore here are the changes I propose in your PR:

- remove the type `dispatchCmd`
- rollback the `dispatch` function as it is on master
- do not execute the reader's `dispatch` in a goroutine

What do you think?

**Comment by asticode** [comment]
cheers

**Pull request merged**